src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.MoveTypeEditor.cs

```diff
@@ -0,0 +1,226 @@
+
+                // collect the parent chain of declarations to keep.
+                SyntaxNode node = typeNode;
+                while (node.Parent != null && !(node.Parent is TCompilationUnitSyntax))
+                {
+                    spine.Add(node.Parent);
+                    node = node.Parent;
+                }
+
+                // get potential namespace, types and members to remove.
+                var removableCandidates = root
+                    .DescendantNodes(n => DescendIntoChildren(n, spine.Contains(n)))
+                    .Where(n => FilterToTopLevelMembers(n, typeNode));
+
+                // diff candidates with items we want to keep.
+                return removableCandidates.Except(spine);
+            }
+
+            private static bool DescendIntoChildren(SyntaxNode node, bool shouldDescendIntoType)
+            {
+                // 1. get top level types and namespaces to remove.
+                // 2. descend into types and get members to remove, only if type is part of spine, which means
+                //    we'll be keeping the type declaration but not other members, in the new file.
+                return node is TCompilationUnitSyntax
+                    || node is TNamespaceDeclarationSyntax
+                    || (node is TTypeDeclarationSyntax && shouldDescendIntoType);
+            }
+
+            private static bool FilterToTopLevelMembers(SyntaxNode node, SyntaxNode typeNode)
+            {
+                // It is a type declaration that is not the node we've moving
+                // or its a container namespace, or a member declaration that is not a type,
+                // thereby ignoring other stuff like statements and identifiers.
+                return node is TTypeDeclarationSyntax
+                    ? !node.Equals(typeNode)
+                    : (node is TNamespaceDeclarationSyntax || node is TMemberDeclarationSyntax);
+            }
+
+            /// <summary>
+            /// if a nested type is being moved, this ensures its containing type is partial.
+            /// </summary>
+            /// <param name="documentEditor">document editor for the new document being created</param>
+            /// <param name="typeNode">type being moved</param>
+            private void AddPartialModifiersToTypeChain(DocumentEditor documentEditor, TTypeDeclarationSyntax typeNode)
+            {
+                var semanticFacts = State.SemanticDocument.Document.GetLanguageService<ISemanticFactsService>();
+                var typeChain = typeNode.Ancestors().OfType<TTypeDeclarationSyntax>();
+
+                foreach (var node in typeChain)
+                {
+                    var symbol = (ITypeSymbol)State.SemanticDocument.SemanticModel.GetDeclaredSymbol(node, CancellationToken);
+                    if (!semanticFacts.IsPartial(symbol))
+                    {
+                        documentEditor.SetModifiers(node, DeclarationModifiers.Partial);
```

Comment:
Did you confirm that 'SetModifiers' works properly when it has existing modifiers?